Comprehensive SVML unit test

anton-malakhov opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->
Unit test covering code-generation for more SVML functions at once. It creates foundation for future extensions covering more functions or documenting these, which will not work in Numba.

Currently, it covers 12 numpy or 20 math functions in 48 modes: (2 types, 3 arch x 2 flags x 4 use-cases). The list of SVML functions not yet supported by the test or not represented in Python include: pow atan2 cbrt cdfnorm cdfnorminv ceil cosd erfcinv erfinv exp10 exp2 floor fmod hypot invsqrt log2 logb nearbyint rint round sind trunc.

anton-malakhov commented:
@stuartarchibald could you please quickly check if this code is compliant to any python/numba/style practices/guidelines I might have missed here. E.g. I'm concerned with lots of TODO marks I left there. Thanks!

stuartarchibald commented on numba/tests/test_svml.py:
```diff
@@ -6,15 +6,172 @@
+            asm = jit.library.get_asm_str()
+            missed = [pattern for pattern in contains if not pattern in asm]
+            found  = [pattern for pattern in avoids if pattern in asm]
+            assert not missed and not found, "While expecting %s,\n\tit contains %s:\n%s\nwhen compiling %s" % \
```
Perhaps use members of the `self.assert*` family to be consistent with other test code.

stuartarchibald commented on numba/tests/test_svml.py:
```diff
@@ -6,15 +6,172 @@
+            found  = [pattern for pattern in avoids if pattern in asm]
+            assert not missed and not found, "While expecting %s,\n\tit contains %s:\n%s\nwhen compiling %s" % \
+                (str(missed), str(found), '\n'.join(
+                [line for line in asm.split('\n') if re.search('\$[a-z_]\w+,', line)]), fn.__doc__)
```
Perhaps hoist the regex to class member level?

stuartarchibald commented:
@anton-malakhov looks good. Style wise, Numba tries to format as PEP8, `autopep8` can do this for you if desirable. I added a couple of comments, but overall the approach looks good. In relation to `TODOs` they are useful left in, perhaps just add a new issue for code that isn't written if it isn't going to make it into this PR? The only other comment I have is to please include plenty of docstrings/comments over what is going on and how to extend the testing in the future. I'll do a full review when you remove the `[WIP]` from the title :) Thanks for the effort!

anton-malakhov commented:
Thank you, @stuartarchibald ! I've addressed your comment and aligned the code with autopep8 suggestions (except few cases of excessive indentation). I also extended it onto `sqrt` function and `float32` type. I think it's good enough for now and the rest of the functions and modes can be covered when we'll have additional motivation/interest for testing these.

stuartarchibald commented:
Thanks for the fixes @anton-malakhov , looks good. Merging.
